lib/statsd/instrument/backends/store_backend.rb

```diff
@@ -0,0 +1,73 @@
+require 'monitor'
+require 'time'
+require 'fileutils'
+
+module StatsD::Instrument::Backends
+  class StoreBackend < StatsD::Instrument::Backend
+    MAXIMUM_NAME_LENGTH = 256
+    TYPE_PACKS = {
+      c:  'Qq',
+      ms: 'Qf',
+      g:  'Qf',
+      h:  'Qf',
+    }
+    MetricNameStruct = Struct.new(:id, :name, :type) do
+      def to_bin
+        raise "Metric name cannot be longer than #{MAXIMUM_NAME_LENGTH}" if name.size > MAXIMUM_NAME_LENGTH
+        [id, name, type.to_s].pack("QA#{MAXIMUM_NAME_LENGTH}A2")
+      end
+    end
+    MetricStruct = Struct.new(:id, :value, :type) do
+      def to_bin
+        raise "Metrid ID must be higher than 0" unless id > 0
+        raise "Type #{type} not supported by store backend" unless pack_str = TYPE_PACKS[type]
+        [id, value].pack(pack_str)
+      end
+    end
+
+    include MonitorMixin
+
+    def initialize(dir_or_io = "/tmp/statsb-temp")
+      super()
+      @dir_or_io = dir_or_io
+      @io = nil
+      reopen
+    end
+
+    def reopen
+      if @dir_or_io.respond_to? :write
+        @io = @dir_or_io
+        @next_reopen = false
+      else
+        @next_reopen = Time.parse("24:00").to_i
+        dirname = Time.now.strftime("%Y-%m-%d")
+        basedir = File.join(@dir_or_io, dirname)
+        FileUtils.mkdir_p basedir
+        @io.close if @io
+        @io = File.open(File.join(basedir, "#{$$}.statsb"), 'ab')
+        @io.write('STATSBv1') if @io.stat.size == 0
+      end
+      @names = {}
+      @next_id = 1
+    end
+
+    def collect_metric(m)
+      if m.sample_rate < 1.0 && rand > m.sample_rate
+        return false
+      end
+
+      reopen if @next_reopen && Time.now.to_i > @next_reopen
```

Comment:
Rather than dealing with rotating log files here, I'd prefer to use a standard unix like `logrotate` to do that for us. That would simplify the code here significantly, and logrotate probably does a better job.